Lib/fontTools/misc/configTools.py

```diff
@@ -0,0 +1,186 @@
+"""
+import logging
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, Iterator, MutableMapping, Union
+
+log = logging.getLogger(__name__)
+
+__all__ = [
+    "ConfigError",
+    "ConfigAlreadyRegisteredError",
+    "ConfigValueParsingError",
+    "ConfigValueValidationError",
+    "ConfigUnknownOptionError",
+    "Option",
+    "OPTIONS",
+    "register_option",
+    "Config",
+]
+
+
+class ConfigError(Exception):
+    """Base exception for the config module."""
+
+
+class ConfigAlreadyRegisteredError(ConfigError):
+    """Raised when a module tries to register a configuration option that
+    already exists.
+
+    Should not be raised too much really, only when developing new fontTools
+    modules.
+    """
+
+    def __init__(self, name):
+        super().__init__(f"Config option {name} is already registered.")
+
+
+class ConfigValueParsingError(ConfigError):
+    """Raised when a configuration value cannot be parsed."""
+
+    def __init__(self, name, value):
+        super().__init__(
+            f"Config option {name}: value cannot be parsed (given {repr(value)})"
+        )
+
+
+class ConfigValueValidationError(ConfigError):
+    """Raised when a configuration value cannot be parsed."""
+
+    def __init__(self, name, value):
+        super().__init__(
+            f"Config option {name}: value is invalid (given {repr(value)})"
+        )
+
+
+class ConfigUnknownOptionError(ConfigError):
+    """Raised when a configuration option is unknown."""
+
+    def __init__(self, name, value):
+        super().__init__(
+            f"Config option {name} is unknown (with given value {repr(value)})"
+        )
+
+
+@dataclass
+class Option:
+    name: str
+    help: str
+    default: Any
+    # Turn input (e.g. string) into proper type. Only when reading from file.
+    parse: Callable
+    # Return true if the given value is an acceptable value
+    validate: Callable
+
+
+OPTIONS: Dict[str, Option] = {}
+
+
+def register_option(
+    name: str, help: str, default: Any, parse: Callable, validate: Callable
+):
+    if name in OPTIONS:
+        raise ConfigAlreadyRegisteredError(name)
+    option = Option(name, help, default, parse, validate)
+    OPTIONS[name] = option
+    return option
+
+
+_USE_GLOBAL_DEFAULT = object()
+
+
+class Config(MutableMapping):
+    _values: Dict
+
+    @classmethod
+    def from_file(cls):
+        """TODO"""
+
+    def __init__(self, values: Union["Config", Dict] = {}, parse_values=False):
+        self._values = {}
+        values_dict = values._values if isinstance(values, Config) else values
+        for name, value in values_dict.items():
+            self.set(name, value, parse_values)
+
+    def set(self, name: str, value: Any, parse_values=False):
+        """
+        Set the value of an option.
+        """
+        try:
+            option = OPTIONS[name]
+        except KeyError:
+            # FIXME: (Jany)  What to do with options we don't recognize?
```

Comment:
could be an init parameter to the Config class, like `skip_unknown=False`

Reply:
Cosimo says: Or maybe a version the config file? Needs more thinking (maybe see what other applications are doing with their config). Maybe set() should always raise if you try to set an unknown option, but when you're parsing a file, maybe there it would make sense to skip unknown options?

Reply:
I added skip_unknown=False